### PR TITLE
Workflow refactor

### PR DESCRIPTION
**Pull request by jlibovicky**
This pull request introduces several important changes in the inner workflow of model execution.

* It introduces `TensorFlowManager`, a class that gathers TensorFlow sessions and their configuration and manages lifecycle of runners — it is now the only place where `session.run` is called.
* Runners are now supposed to be immutable objects, which are able to generate mutable `Executable` which can be executed in the execution loop of the TensorFlow manager.
* Executables have two hooks: in the first one, they tell what should be executed in the TF sessions, in the second one, they collect what was returned from the `session.run` method.
* Yes, writing runners became more complicated, the benefit though should be being able to have multiple runners at the same moment.
* Evaluation configuration has been modified: when we can have multiple runners, we need to specify what series should be evaluated (against which dataset series) and by which function
* The are two runners implemented right now: a greedy runner (which does naive ensembles) and a recurrent runner, which can do a beam search and does the ensembles correctly (and is two times slower that greedy decoding even with the beam size 1)
* Trainer is a special case of runner. There is one entirely general trainer and one special case which our previously known `CrossEntropyTrainer`

There are minor issues know to me:
* collecting and outputing TensorBoard summaries may not work correctly
* the beam search and ensemble can be made faster by using `partial_run` in TensorFlowManager
* perplexity runner does not follow this new paradigm
* scoring in beamsearch should be done dynamically, not computed from scracth always

Please have a look at the code, made me comment everything you don't understand.

@tomasmcz: Asi to bude nějakou dobu trvat, než se to zreviewuje a zmerguje, ale když na to tak koukám, tak nemyslím, že by tam bylo moc věcí, co by bránily začít s předěláváním té konfigurace.

### CODE REVIEW

File: neuralmonkey/runners/rnn_runner.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    n_best_indices = unsorted_n_best_indices[
+        np.argsort(scores[unsorted_n_best_indices])]
+    return n_best_indices
+
+
+def _score_expanded(n: int,
+                    batch_size: int,
+                    expanded: List[ExpandedBeamBatch],
+                    scoring_function) -> \
+        Tuple[List[np.ndarray], List[np.ndarray]]:
```
Comment (jindrahelcl) [inline neuralmonkey/runners/rnn_runner.py]: A docstring would help here.

File: neuralmonkey/runners/runner.py [hunk 2]
```diff
@@ -1,50 +1,69 @@
...
+    def __init__(self, output_series, decoder):
+        super(GreedyRunner, self).__init__(output_series, decoder)
+
+    def get_executable(self, train=False, summaries=True):
+        if train:
+            losses = [self._decoder.train_loss,
+                      self._decoder.runtime_loss]
+        else:
+            losses = [tf.zeros([]), tf.zeros([])]
+        to_run = losses + self._decoder.runtime_logprobs
```
Comment (jindrahelcl) [inline neuralmonkey/runners/runner.py]: to počítání nul je blbost, stejně jako pak to počítání těch indexů do toho pole výsledků.
Jako ``fetches`` může ``session.run`` dostat libovolně strukturovanou pythoní věc, takže i dict. Stačí to udělat takhle:

```
fetches = {"logprobs": self._decoder.runtime_logprobs}

if train:
    fethes["losses"] = {"train": self._decoder.train_loss,
                        "runtime": self._decoder.runtime_loss}

return GreedyRunExecutable(self.all_coders, fetches, self._decoder.vocabulary)
```

Jestli teda ``fetches`` nikde někdo nijak jinak nepoužívá, než že to cpe do ``session.run``

File: neuralmonkey/runners/runner.py [hunk 3]
```diff
@@ -1,50 +1,69 @@
...
+        """Get the feedables and tensors to run."""
+        return self.all_coders, self.to_run, {}
+
+    def collect_results(self, results: List[List[RunResult]]) -> None:
+        train_loss = 0.
+        runtime_loss = 0.
+        summed_logprobs = [-np.inf for _ in range(len(self.to_run) - 2)]
+
+        for sess_result in results:
+            train_loss += sess_result[0]
```
Comment (jindrahelcl) [inline neuralmonkey/runners/runner.py]: Tady pak bude jen:

```
train_loss = sess_result["losses"]["train"]
runtime_loss = sess_result["losses"]["runtime"]
logprobs = sess_result["logprobs"]
```

File: neuralmonkey/runners/runner.py [hunk 4]
```diff
@@ -1,50 +1,69 @@
...
+            runtime_loss += sess_result[1]
+
+            for i, logprob in enumerate(sess_result[2:]):
+                summed_logprobs[i] = np.logaddexp(summed_logprobs[i], logprob)
```
Comment (jindrahelcl) [inline neuralmonkey/runners/runner.py]: k čemu je přesně tohle?

File: neuralmonkey/tf_manager.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                                .format(len(variable_files), num_sessions))
+            self.restore(variable_files)
+
+    # pylint: disable=too-many-locals
+    def execute(self,
+                dataset: Dataset,
+                execution_scripts,
+                train=False,
+                summaries=True,
+                batch_size=None) -> List[ExecutionResult]:
```
Comment (jindrahelcl) [inline neuralmonkey/tf_manager.py]: type hints

File: neuralmonkey/tf_manager.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                additional_feed_dicts = []
+                tensor_list_lengths = []  # type: List[int]
+
+                for executable in executables:
+                    if executable.result is None:
+                        (feedables,
+                         tensors_to_execute,
+                         add_feed_dict) = executable.next_to_execute()
+                        all_feedables = all_feedables.union(feedables)
+                        all_tensors_to_execute.extend(tensors_to_execute)
```
Comment (jindrahelcl) [inline neuralmonkey/tf_manager.py]: Tady vidim jedinou záškobrtku v tom způsobu fetchování popsanym v runneru.. Ale asi bude stačit tady vyměnit extend za update a pak asi ještě šéfovat, aby nekolidovaly klíče v těch dictionaries.. Ale on by vlastně ``all_tensors_to_execute`` mohl být dict indexovanej runner objektama a pak tohle by ani nebylo update, ale jen přiřazení.

A úplně se ten kód zbaví tý režie okolo, jako ``tensor_list_lengths``, nebo k čemu je ta proměnná.

File: neuralmonkey/tf_manager.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+        for sess, file_name in zip(self.sessions, variable_files):
+            log("Loading variables from {}".format(file_name))
+            self.saver.restore(sess, file_name)
+
+
+def _partition_results(session_results: List[List[RunResult]],
+                       tensor_list_lengths: List[int]) \
+                                                 -> List[List[List[RunResult]]]:
+    """Split the session run results back for their executables."""
```
Comment (jindrahelcl) [inline neuralmonkey/tf_manager.py]: Tahle funkce pak úplně zmizí

File: neuralmonkey/trainers/generic_trainer.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+        if optimizer is None:
+            self.optimizer = tf.train.AdamOptimizer(1e-4)
+        else:
+            self.optimizer = optimizer
+
+        with tf.variable_scope('regularization'):
+            regularizable = [tf.reduce_sum(
+                v ** 2) for v in tf.trainable_variables()
+                             if BIAS_REGEX.findall(v.name)]
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: fuj!

použít místo toho ``regularizer`` parametr funkce ``tf.get_variable()`` a pak vyzobat loss z kolekce ``GraphKeys.REGULARIZATION_LOSSES``

Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: Ale nefixovat to v tomhle PR

File: neuralmonkey/trainers/generic_trainer.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        self.all_coders = all_coders
+        self.train_op = train_op
+        self.losses = losses
+        self.scalar_summaries = scalar_summaries
+        self.histogram_summaries = histogram_summaries
+
+        self.result = None
+
+    def next_to_execute(self) -> NextExecute:
+        fetches = [self.train_op]
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: Vylepšit fetches strukturu

File: scripts/postedit_prepare_data.py [hunk 10]
```diff
@@ -17,56 +17,67 @@
 # tests: lint
```
Comment (jindrahelcl) [inline scripts/postedit_prepare_data.py]: tohle jsou jen refactory nebo je tu nějaká změna?

File: neuralmonkey/tf_manager.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+            if len(variable_files) != num_sessions:
+                raise Exception(("The number of provided variable files ({}) "
+                                 "is different than a number sessions ({})")
+                                .format(len(variable_files), num_sessions))
+            self.restore(variable_files)
+
+    # pylint: disable=too-many-locals
+    def execute(self,
+                dataset: Dataset,
+                execution_scripts,
```
Comment (jindrahelcl) [inline neuralmonkey/tf_manager.py]: Co jsou execution_scripts? To jsou runnery?

File: neuralmonkey/trainers/generic_trainer.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+# pylint: disable=invalid-name
+Gradients = List[Tuple[tf.Tensor, tf.Variable]]
+Objective = NamedTuple('Objective',
+                       [('name', str),
+                        ('decoder', Any),
+                        ('loss', tf.Tensor),
+                        ('gradients', Optional[Gradients])])
+
+BIAS_REGEX = re.compile(r'[Bb]ias')
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: Navíc ten regex ani nepojme všechno. obcas  ty biasy končej na ``_b``

File: neuralmonkey/learning_utils.py [hunk 13]
```diff
@@ -1,172 +1,68 @@
...
+                  train_dataset: Dataset,
+                  val_dataset: Dataset,
+                  log_directory: str,
+                  evaluators: EvalConfiguration,
+                  runners: List[BaseRunner],
+                  test_datasets=None,
                   save_n_best_vars=1,
                   link_best_vars="/tmp/variables.data.best",
                   vars_prefix="/tmp/variables.data",
-                  initial_variables=None,
```
Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: `initial_variables` jsou teď vlastnost TensorFlowManageru.

File: neuralmonkey/learning_utils.py [hunk 14]
```diff
@@ -298,192 +188,228 @@ def argworst(scores, minimize):
...
+                               eval_result: Evaluation,
+                               seen_instances: int,
+                               execution_results: List[ExecutionResult],
+                               train=False) -> None:
+    """Log the evaluation results and the TensorBoard summaries."""
 
     if train:
-        color = 'yellow'
-        prefix = 'train'
+        color, prefix = 'yellow', 'train'
```
Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Dalo :-)

File: neuralmonkey/runners/base_runner.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+    def get_executable(self, train=False, summaries=True) -> Executable:
+        raise NotImplementedError()
+
+    @property
+    def loss_names(self) -> List[str]:
+        raise NotImplementedError()
+
+def reduce_execution_results(execution_results: List[ExecutionResult]) \
+                                                           -> ExecutionResult:
```
Comment (jlibovicky) [inline neuralmonkey/runners/base_runner.py]: Jde.

File: neuralmonkey/tf_manager.py (hunk 11 above)
Comment (jlibovicky) [inline neuralmonkey/tf_manager.py]: Jo, to jsou ty runnery. Chtělo by to vymyslet nějaký vhodnější ší hyperonymum pro runnery a trainery.

File: neuralmonkey/trainers/generic_trainer.py (hunk 12 above)
Comment (jlibovicky) [inline neuralmonkey/trainers/generic_trainer.py]: Stejně se to brzo předělá, ne?

File: neuralmonkey/tf_manager.py (hunk 11 above)
Comment (jindrahelcl) [inline neuralmonkey/tf_manager.py]: co třeba executor? :-)

File: neuralmonkey/learning_utils.py (hunk 13 above)
Comment (jindrahelcl) [inline neuralmonkey/learning_utils.py]: už to vidim

File: neuralmonkey/runners/base_runner.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    """Collect recusively all encoders and decoders."""
+    if hasattr(coder, 'encoders'):
+        return set([coder]).union(*(collect_encoders(enc)
+                                    for enc in coder.encoders))
+    else:
+        return set([coder])
+
+
+class BaseRunner(object):
+    def __init__(self, output_series: str, decoder) -> None:
```
Comment (jindrahelcl) [inline neuralmonkey/runners/base_runner.py]: Aspoň Any, nebo tak něco.. (?) Souvisí to s mojí otázkou nahoře, co je víc pythonický..

File: neuralmonkey/runners/runner.py (hunk 3 above)
Comment (jlibovicky) [inline neuralmonkey/runners/runner.py]: A to se může takhle více-úrovňovně zanořovat slovník?

File: neuralmonkey/runners/runner.py (hunk 4 above)
Comment (jlibovicky) [inline neuralmonkey/runners/runner.py]: To je enseble: vezme to výsledky z více sessions a sečte v logaritmický doméně.

**Comment by jindrahelcl** [comment]
Co se týče zalamování těch importů backslashama, dočet jsem se, že se taky daj dávat do závorek:
https://www.python.org/dev/peps/pep-0328/#id7

File: neuralmonkey/runners/runner.py (hunk 3 above)
Comment (jindrahelcl) [inline neuralmonkey/runners/runner.py]: Jestli se ptáš obecně v Pythonu, tak jsem teď zkusil napsat 

```
a = {"b": {"foo": 5}}
print(a["b"]["foo"])
```

A napsalo to 5

File: neuralmonkey/tf_manager.py (hunk 11 above)
Comment (jindrahelcl) [inline neuralmonkey/tf_manager.py]: Nevim... tak možná bych napsal runners, s tim, že začnem brát trainer jako speciální druh runneru.

File: neuralmonkey/runners/runner.py (hunk 3 above)
Comment (jlibovicky) [inline neuralmonkey/runners/runner.py]: Myslel jsem tohle:
```
In [7]: sess.run({"haf": {"baf": x + 1}}, feed_dict={x : 0})
Out[7]: {'haf': {'baf': 1}}
```
A to taky funguje. :-)

Comment (jindrahelcl) [inline neuralmonkey/runners/runner.py]: Jo, jo, do session.run můžeš dát libovolně strukturovanej diktolist:

> The fetches argument may be a single graph element, or an arbitrarily nested list, tuple, namedtuple, or dict containing graph elements at its leaves.

File: neuralmonkey/runners/base_runner.py (hunk 16 above)
Comment (jlibovicky) [inline neuralmonkey/runners/base_runner.py]: No to já nevim, psát ty anotace je IMHO docela exotický samo o sobě.

File: neuralmonkey/tf_manager.py (hunk 11 above)
Comment (jlibovicky) [inline neuralmonkey/tf_manager.py]: Co kdyby se předek runnerů a trainerů jmenoval `Executable` a to, co je teď  `Executable` by se jmenovala `ExePackage`?

**Comment by jlibovicky** [comment]
Hm, to jsou ty dicty dictů, co se tam teď posílají do session.run. S tím asi nejde udělat nic, než všechny tyhle věci změnit na generickej dict.

**Comment by jlibovicky** [comment]
A mělo by to být všechno.

**Comment by jindrahelcl** [comment]
Ok.
